sbi/neural_nets/embedding_nets/resnet.py

```diff
@@ -0,0 +1,321 @@
+
+        # Preserve the input dimensionality if the output is not specified
+        if c_out is None:
+            c_out = c_in
+
+        # Preserve the dimensionality of the input and the output
+        if c_in == c_out:
+            self.residual = nn.Identity()
+
+        # Apply 1x1 convolutions to change the number of channels
+        elif (c_in != c_out) and change_c_mode == "conv":
+            self.residual = nn.Conv2d(c_in, c_out, kernel_size=1, stride=1, bias=False)
+
+        # Fill the required additional channels with zeros
+        elif (c_in != c_out) and change_c_mode == "zeros":
+            # Checkt if there are more output channels than input channels
+            if c_in > c_out:
+                raise ValueError("c_in channels must be smaller than c_out.")
+
+            self.residual = partial(zero_pad, c_out=c_out)
+
+        ###############################################################################
+        # Halve the height and width of the input
+        ###############################################################################
+
+        # Apply a strided convolution if the input is downsampled
+        if downsample:
+            self.downsampling = nn.Conv2d(
+                c_out, c_out, kernel_size=3, stride=2, bias=False, padding=1
+            )
+
+        # No downsampling
+        else:
+            self.downsampling = nn.Identity()
+
+        ###############################################################################
+        # Define the transformation of the residual block
+        ###############################################################################
+
+        self.f = construct_mapping(
+            c_in, c_out, downsample=downsample, **construct_mapping_kwargs
+        )
+
+        self.activation = activation()
+
+    def forward(self, x: torch.tensor) -> torch.tensor:
+        """
+        Forward pass of the residual block.
+
+        Args:
+            x:      Input tensor
+
+        Returns:
+            y:      Output tensor
+        """
+
+        # Check if image like data is used
+        assert len(x.shape) == 4
+
+        # Compute the skip connection
+        r = self.downsampling(self.residual(x))
+
+        # Full transformation
+        y = self.activation(self.f(x) + r)
+
+        return y
+
+
+class ResNetEmbedding(nn.Module):
+    def __init__(
+        self,
+        c_in: int,
+        c_out: int = 20,
+        c_hidden_fc: int = 1000,
+        n_stages: int = 4,
+        blocks_per_stage: list | None = None,
+        c_stages: list | None = None,
+        activation: Type[nn.Module] = nn.ReLU,
+        change_c_mode: str = "conv",
+        construct_mapping: Callable = construct_simple_conv_net,
+        construct_mapping_kwargs: dict | None = None,
+        coupling_block_kwargs: dict | None = None,
+    ) -> None:
+        """
+        Residual neural network mapping image like data to a fixed sized vector.
```

Comment:
After the first introductory sentence, it is good for user-facing elements to include a more detailed description of this particular embedding net. This can be small amouts of backgroudn but also how to use it i.e. what input shapes are expected and so on.